src/System.Console/src/System/IO/StdInStreamReader.cs

```diff
@@ -0,0 +1,259 @@
+using System.IO;
+using System.Text;
+using System.Runtime.InteropServices;
+
+namespace System.IO
+{
+    /* This class is used by for reading from the stdin.
+     * It is designed to read stdin in raw mode for interpreting
+     * key press events and maintain its own buffer for the same.
+     * which is then used for all the Read operations
+     */
+    internal class StdInStreamReader : StreamReader
+    {
+        private char[] _unprocessedBufferToBeRead; // Buffer that might have already been read from stdin but not yet processed.
+        private const int BytesToBeRead = 255; // No. of bytes to be read from the stream at a time.
+        private int _startIndex; // First unprocessed index in the buffer;
+        private int _endIndex; // Last unprocessed index in the buffer;
+        private StringBuilder _readLineSB;
+
+        internal StdInStreamReader(Stream stream, Encoding encoding, int bufferSize) : base(stream: stream, encoding: encoding, detectEncodingFromByteOrderMarks: false, bufferSize: bufferSize, leaveOpen: true)
+        {
+            _unprocessedBufferToBeRead = new char[encoding.GetMaxCharCount(bufferSize)];
+            _startIndex = _endIndex = -1; // There is no unprocessed index yet.
+            _readLineSB = new StringBuilder();
+        }
+
+        /// <summary> Checks whether the unprocessed buffer is empty. </summary>
+        bool IsExtraBufferEmpty()
+        {
+            if (_startIndex == -1 && _endIndex == -1) return true;
+
+            if (_startIndex > _endIndex) // Everything has been processed;
+            {
+                _startIndex = -1;
+                _endIndex = -1;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary> Reads the buffer in the raw mode. </summary>
+        unsafe void ReadExtraBuffer()
+        {
+            Debug.Assert(IsExtraBufferEmpty());
+
+            byte* bufPtr = stackalloc byte[BytesToBeRead];
+            int result;
+            Interop.CheckIo(result = Interop.Sys.ReadStdinUnbuffered(bufPtr, BytesToBeRead));
+
+            Debug.Assert(result <= BytesToBeRead);
+
+            // Now we need to convert the byte buffer to char buffer.
+            fixed (char* unprocessedBufPtr = _unprocessedBufferToBeRead)
+            {
+                _startIndex = 0;
+                int charCount = CurrentEncoding.GetChars(bufPtr, result, unprocessedBufPtr, _unprocessedBufferToBeRead.Length);
+                _endIndex = charCount - 1;
+            }
+        }
+
+        public override string ReadLine()
+        {
+            //Check if there is anything left in the _unprocessedBufferToBeRead.
+            while (!IsExtraBufferEmpty())
+            {
+                ConsoleKeyInfo keyInfo = ReadKey();
+
+                if (keyInfo.Key == ConsoleKey.Enter)
+                    return _readLineSB.ToString();
+
+                else if (keyInfo.Key == ConsoleKey.Backspace)
+                {
```

Comment:
Nit: if we have braces on any of the bodies in an if / else, I suggest we have them on all parts of it.